src/outpack/location.py

```diff
@@ -71,6 +82,149 @@ def outpack_location_rename(old, new, root=None, *, locate=True):
     update_config(config, root.path)
 
 
+def location_resolve_valid(
+    location, root, include_local, include_orphan, allow_no_locations
+):
+    if location is None:
+        location = outpack_location_list(root)
+    elif isinstance(location, str):
+        if location not in outpack_location_list(root):
+            msg = f"Unknown location: '{location}'"
+            raise Exception(msg)
+        else:
+            location = [location]
+    elif isinstance(location, collections.abc.Iterable) and all(
+        isinstance(item, str) for item in location
+    ):
+        unknown = set(location).difference(outpack_location_list(root))
```

Comment:
Not sure if this is clearer?

```suggestion
        unknown = set(location) - set(outpack_location_list(root))
```

I wonder if a utility function that does

```python
def setdiff(a, b):
  return set(a) - set(b)
```

or even

```python
def setdiff(a, b):
  return type(a)(set(a) - set(b))
```

I think we'll do this a lot, and wonder if this would be easier to read. We should be careful that when porting code over from the R uses of this though that in a few places I've exploited the sorting properties of that function, which would then mean that we'd be looking for:

```
python
def setdiff(a, b):
  ret = []
  for el in a:
    if el not in b and el not in ret:
      ret.add(el)
  return ret
```

which is slower. Or add a following sorting phase to the previous approach

Reply:
Bit of docs I found https://docs.python.org/2/library/stdtypes.html#set-types-set-frozenset suggests preferring the non-operator versions

> Note, the non-operator versions of [union()](https://docs.python.org/2/library/stdtypes.html#frozenset.union), [intersection()](https://docs.python.org/2/library/stdtypes.html#frozenset.intersection), [difference()](https://docs.python.org/2/library/stdtypes.html#frozenset.difference), and [symmetric_difference()](https://docs.python.org/2/library/stdtypes.html#frozenset.symmetric_difference), [issubset()](https://docs.python.org/2/library/stdtypes.html#frozenset.issubset), and [issuperset()](https://docs.python.org/2/library/stdtypes.html#frozenset.issuperset) methods will accept any iterable as an argument. In contrast, their operator based counterparts require their arguments to be sets. This precludes error-prone constructions like set('abc') & 'cbs' in favor of the more readable set('abc').intersection('cbs').

Could add a helper though with optional arg for `preserve_order` or something though? But I think `a.intersect(b)` is quite readable?